cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc.cpp

```diff
@@ -96,65 +95,153 @@ struct train_kernel_gpu<Float, method::lloyd_dense, task::clustering> {
-                                                               daal_initial_centroids.get() };
-
-        daal::data_management::NumericTable* daal_output[4] = { daal_centroids.get(),
-                                                                daal_labels.get(),
-                                                                daal_objective_function_value.get(),
-                                                                daal_iteration_count.get() };
-        interop::status_to_exception(
-            daal_kmeans_lloyd_dense_ucapi_kernel_t<Float>().compute(daal_input, daal_output, &par));
-
-        auto [arr_iter_count_host, arr_iter_count_event] = dal::backend::to_host(arr_iter_count);
-        auto [arr_obj_func_host, arr_obj_func_event] = dal::backend::to_host(arr_obj_func);
-        sycl::event::wait_and_throw({ arr_iter_count_event, arr_obj_func_event });
 
+        std::int64_t block_rows = get_block_size_in_rows<Float>(queue, column_count);
+        std::int64_t part_count =
+            get_part_count_for_partial_centroids<Float>(queue, column_count, cluster_count);
+
+        auto arr_distance_block = pr::ndarray<Float, 2>::empty(queue,
+                                                               { block_rows, cluster_count },
+                                                               sycl::usm::alloc::device);
+        auto arr_centroids = pr::ndarray<Float, 2>::empty(queue,
+                                                          { cluster_count, column_count },
+                                                          sycl::usm::alloc::device);
+        auto arr_partial_centroids =
+            pr::ndarray<Float, 2>::empty(queue,
+                                         { part_count * cluster_count, column_count },
+                                         sycl::usm::alloc::device);
+        auto arr_labels =
+            pr::ndarray<std::int32_t, 2>::empty(queue, { row_count, 1 }, sycl::usm::alloc::device);
+        auto arr_closest_distances =
+            pr::ndarray<Float, 2>::empty(queue, { row_count, 1 }, sycl::usm::alloc::device);
+        auto arr_counters =
+            pr::ndarray<std::int32_t, 1>::empty(queue, cluster_count, sycl::usm::alloc::device);
+        auto arr_candidate_indices =
+            pr::ndarray<std::int32_t, 1>::empty(queue, cluster_count, sycl::usm::alloc::device);
+        auto arr_candidate_distances =
+            pr::ndarray<Float, 1>::empty(queue, cluster_count, sycl::usm::alloc::device);
+        auto arr_num_empty_clusters =
+            pr::ndarray<std::int32_t, 1>::empty(queue, 1, sycl::usm::alloc::device);
+        auto arr_objective_function =
+            pr::ndarray<Float, 1>::empty(queue, 1, sycl::usm::alloc::device);
+
+        Float prev_objective_function = det::limits<Float>::max();
+        std::int64_t iter;
+        sycl::event centroids_event;
+
+        for (iter = 0; iter < max_iteration_count; iter++) {
```

Comment:
Can we simplify body of this loop using abstractions? For example, introduce class called `cluster_updater` and call `updater.update(...)`.

Reply:
Done